include/ozo/binary_deserialization.h

```diff
@@ -232,4 +232,23 @@ basic_result<T>& recv_result(basic_result<T>& in, const oid_map_t<M>&, basic_res
     return out;
 }
 
+template <typename T, typename M>
+basic_result<T>& recv_result(basic_result<T>& in, const oid_map_t<M>& oid_map, std::reference_wrapper<basic_result<T>> out) {
```

Comment:
This is strange, we already use basic_result as an output argument which being accepted by non-constant reference only, what is the point to support a reference wrapper for reference argument?

Reply:
With this commit you can't anymore. `async_request_out_handler` shouldn't store raw reference. For example if you pass `std::shared_ptr` to `make_async_request_op` before this commit, you will have only reference to it, not a copy. This example shows the problem:
```c++
ozo::basic_result<std::shared_ptr<PGresult>> res(
    std::shared_ptr<PGresult>(reinterpret_cast<PGresult*>(1), [] (auto& v) { v = 0; })
);

EXPECT_EQ(res.handle().get(), reinterpret_cast<PGresult*>(1));

auto op = ozo::impl::make_async_request_op("SELECT 1"_SQL + " + 1"_SQL, res, [] (auto ...) {});

res.handle().reset();

EXPECT_EQ(op.out_.out.handle().get(), nullptr);
// EXPECT_EQ(op.out_.out.handle().get(), reinterpret_cast<PGresult*>(1)); // desired behaviour
```

My opinion is to force users to specify whether they want to copy result or to pass reference.

Reply:
But what for? It must be a non-const reference (or rvalue reference if we construct inserter right on the function call) always because the user is responsible for the output object's lifetime and condition. In your example it is going something strange - the user is trying to destroy the output object on purpose. Why do you want something like a copy of std::shared_ptr instead of reference. I mean what is the aim of this change? It is not clear from the example. Are there more reasonable things like dangling references inside the library or something like this, but without user's manipulations with the output object?

> My opinion is to force users to specify whether they want to copy result or to pass reference.

User do not want to copy result, because result is the output and the only way to get output is to have original result object or reference to. There is just no option at all.

Reply:
> the user is trying to destroy the output object on purpose

If you allocate `res` on a stack and don't copy anywhere else same thing is happen.

If I do not capture res into lambda, I have SIGSEGV. It is not obvious not to copy shared_ptr, but take only a reference.
```c++
io.post([&] {
    ozo::shared_result res;
    ozo::request(ozo::make_connector(conn_info, io), "SELECT 1"_SQL, res,
        [] (auto, auto) {});
});
```

Another example. I pass iterator using lvalue, and surprise is request takes only reference to it. So I have SIGSEGV.
```c++
rows_of<std::vector<std::string>> res(1);
io.post([&] {
    auto it = res.begin();
    ozo::request(ozo::make_connector(conn_info, io), "SELECT "_SQL + foos, it,
            [](auto, auto) {});
});
```

Reply:
Well, this is the behavior we define via the interface. A user has to control the lifetime of the output object. That's it. Via such behavior, we do not obligate user to use shared pointers or something like this in case if it is not needed(e.g. Boost.Coroutine or `std::future` or just to make concurrent requests in synchronous code and do not start a different thread for `io.run()`).  If a user decides to use a callback-style asynchronous code, then he or she has to control objects lifetime, as usual. This is common practice.